### ISSUE TITLE
Fix for `clippy::cmp_owned` on `PathBuf::from` causes compilation error

### ISSUE THREAD

**Issue by GideonBear** [rust-lang/rust-clippy#16448]
### Summary

See reproduction.
Can reproduce on both stable and nightly.

### Reproducer

Code:

```rust
use std::path::PathBuf;


fn main() {
    let a = PathBuf::from("some/path");
    if a == PathBuf::from("some/path") {
        println!("Expected");
    }
}
```

Current output:
```console
$ cargo clippy --fix --allow-dirty
    Checking delme v0.1.0 (/home/gideon/delme)
warning: failed to automatically apply fixes suggested by rustc to crate `delme`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0277]: can't compare `std::path::PathBuf` with `&str`
 --> src/main.rs:6:10
  |
6 |     if a == "some/path" {
  |          ^^ no implementation for `std::path::PathBuf == &str`
  |
  = help: the trait `std::cmp::PartialEq<&str>` is not implemented for `std::path::PathBuf`
help: consider dereferencing both sides of the expression
  |
6 |     if *a == *"some/path" {
  |        +     +

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0277`.
Original diagnostics will follow.

warning: this creates an owned instance just for comparison
 --> src/main.rs:6:13
  |
6 |     if a == PathBuf::from("some/path") {
  |        -----^^^^^^^^^^^^^^^^^^^^^^^^^^
  |        |
  |        help: try: `a == "some/path"`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#cmp_owned
  = note: `#[warn(clippy::cmp_owned)]` on by default

warning: `delme` (bin "delme") generated 1 warning (run `cargo clippy --fix --bin "delme" -p delme` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.08s
```

Desired output:
Either a good fix or no fix :)

### Version

```text
rustc 1.94.0-nightly (fe98ddcfc 2026-01-17)
binary: rustc
commit-hash: fe98ddcfcfb6f185dbf4adeaf439d8a756da0273
commit-date: 2026-01-17
host: x86_64-unknown-linux-gnu
release: 1.94.0-nightly
LLVM version: 21.1.8
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by GideonBear** [comment]
Looks like the proposed fix (`dereferencing both sides of the expression`) does make it compile, although I don't really understand why (what even is `*"string"`, I thought a `str` couldn't exist?)

**Comment by GideonBear** [comment]
And why can't we `impl PartialEq<&str> for PathBuf`?

**Comment by Jarcho** [comment]
Why can't `str` exist? It's a perfectly valid type, just one that isn't sized. Note `lhs == rhs` is effectively `<typeof(lhs) as PartialEq<typeof(rhs)>::eq(&lhs, &rhs)` which means with both derefs you get `<Path as PartialEq<str>>::eq(&*a, &*"some/path")`.

I don't think anything stops `impl PartialEq<&str> for PathBuf` it just hasn't been done.

**Comment by GideonBear** [comment]
Thanks for the explanation!

> I don't think anything stops `impl PartialEq<&str> for PathBuf` it just hasn't been done.

Then do we want to add that, or change the fix?
